### PR TITLE
Added kafka message timestamp in Message class

### PR DESCRIPTION
**Pull request by tygrash**
A quick question for you Henry - Does Kafka FetchRequest and FetchResponse apis support extraction of kafka message timestamp? 

Acc. to my understanding, it was used till Kafka 0.8.0 version and until then no kafka message timestamp was there. 

Should I update code in KafkaClient.java according to new libraries?

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -56,9 +56,9 @@ In most of the cases, the output of the copy backup is not suitable for direct c
 
 ## Partition group
 
-Partition group clusters log messages based on custom logic parametrized with message content. More precisely, a consumer in this group parses the message content to extract a list of partitions and places the message in a file with other messages that belong to those partitions. In the majority of cases, the message partition list contains a single value, the date corresponding to the timestamp recorded in the message.
+Partition group clusters log messages based on custom logic parametrized with message content. More precisely, a consumer in this group parses the message content to extract a list of partitions and places the message in a file with other messages that belong to those partitions. In the majority of cases, the message partition list contains a single value, the date corresponding to the mTimestamp recorded in the message.
```
Comment (HenryCaiHaiying) [inline DESIGN.md]: Not sure why you need to change the wording from 'timestamp' to 'mTimestamp'.  mTimestamp was just a internal variable name we used for the object and it has an archaic C++ origin, doesn't seem you need to expose the internal object details.

File: src/main/java/com/pinterest/secor/message/Message.java [hunk 2]
```diff
@@ -34,21 +34,23 @@
...
+    private Long mTimestamp;
 
     protected String fieldsToString() {
         return "topic='" + mTopic + '\'' +
                ", kafkaPartition=" + mKafkaPartition +
                ", offset=" + mOffset +
                ", kafkaKey=" + new String(mKafkaKey) +
-               ", payload=" + new String(mPayload);
+               ", payload=" + new String(mPayload) +
+               ", mTimestamp=" + mTimestamp;
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/message/Message.java]: I don't think you need to use 'mTimestamp' in the printout, the other internal variables (payload, kafkaKey) didn't do that either.

File: src/main/thrift/secor.thrift [hunk 3]
```diff
@@ -8,7 +8,7 @@ enum TestEnum {
 }
 
 struct TestMessage {
-    1: required i64 timestamp,
+    1: required i64 mTimestamp,
```
Comment (HenryCaiHaiying) [inline src/main/thrift/secor.thrift]: Why change this variable name?

**Comment by HenryCaiHaiying** [comment]
I didn't quite get why you need to change timestamp to mTimestamp everywhere.

And I don't see the code to extract timestamp from kafka message.

File: DESIGN.md (hunk 1 above)
Comment (tygrash) [inline DESIGN.md]: DESIGN.md and README.md were changed automatically when I ran package command on local. Will revert it in next commit.

File: src/main/java/com/pinterest/secor/message/Message.java (hunk 2 above)
Comment (tygrash) [inline src/main/java/com/pinterest/secor/message/Message.java]: Corrected in next commit. Missed the pattern here. My bad.

File: DESIGN.md (hunk 1 above)
Comment (tygrash) [inline DESIGN.md]: resolved

File: src/main/java/com/pinterest/secor/message/Message.java (hunk 2 above)
Comment (tygrash) [inline src/main/java/com/pinterest/secor/message/Message.java]: resolved

File: src/main/thrift/secor.thrift (hunk 3 above)
Comment (tygrash) [inline src/main/thrift/secor.thrift]: reverted

**Comment by HenryCaiHaiying** [comment]
Are you still going to name the configs: mTimestamp?

**Comment by tygrash** [comment]
Fixed it sir :)

File: pom.xml [hunk 4]
```diff
@@ -621,6 +621,9 @@
         </profile>
         <profile>
             <id>kafka-0.10-dev</id>
+            <activation>
```
Comment (HenryCaiHaiying) [inline pom.xml]: This is probably better to be done in a separate PR, either before and after your PR.  I would prefer to track the version upgrade as its own PR with its own version/tag.

File: src/main/config/secor.common.properties [hunk 5]
```diff
@@ -299,8 +299,6 @@ message.timestamp.name=timestamp
 
 # Separator for defining message.timestamp.name in a nested structure. E.g.
 # {"meta_data": {"created": "1405911096123", "last_modified": "1405912096123"}, "data": "test"}
-# message.timestamp.name=meta_data.created
```
Comment (HenryCaiHaiying) [inline src/main/config/secor.common.properties]: Why these two lines are gone?

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 6]
```diff
@@ -5,9 +5,9 @@
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
  * the License.  You may obtain a copy of the License at
- *
- *    http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p>
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: These header lines might be standard for apache license.  Does Apache publish the new lines?

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 7]
```diff
@@ -61,9 +61,9 @@ private HostAndPort findLeader(TopicPartition topicPartition) {
         try {
             LOG.debug("looking up leader for topic {} partition {}", topicPartition.getTopic(), topicPartition.getPartition());
             consumer = createConsumer(
-                mConfig.getKafkaSeedBrokerHost(),
-                mConfig.getKafkaSeedBrokerPort(),
-                "leaderLookup");
+                    mConfig.getKafkaSeedBrokerHost(),
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: Why do you have those formatting changes?

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 8]
```diff
@@ -98,8 +98,8 @@ private long findLastOffset(TopicPartition topicPartition, SimpleConsumer consum
                 kafka.api.OffsetRequest.LatestTime(), 1));
         final String clientName = getClientName(topicPartition);
         OffsetRequest request = new OffsetRequest(requestInfo,
-                                                  kafka.api.OffsetRequest.CurrentVersion(),
-                                                  clientName);
+                kafka.api.OffsetRequest.CurrentVersion(),
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: ??

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 9]
```diff
@@ -111,24 +111,23 @@ private long findLastOffset(TopicPartition topicPartition, SimpleConsumer consum
         return offsets[0] - 1;
     }
 
-    private Message getMessage(TopicPartition topicPartition, long offset,
-                               SimpleConsumer consumer) {
+    private Message getMessage(TopicPartition topicPartition, long offset, SimpleConsumer consumer) {
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: ??

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 10]
```diff
@@ -111,24 +111,23 @@ private long findLastOffset(TopicPartition topicPartition, SimpleConsumer consum
...
         LOG.debug("fetching message topic {} partition {} offset {}",
                 topicPartition.getTopic(), topicPartition.getPartition(), offset);
         final int MAX_MESSAGE_SIZE_BYTES = mConfig.getMaxMessageSizeBytes();
         final String clientName = getClientName(topicPartition);
         kafka.api.FetchRequest request = new FetchRequestBuilder().clientId(clientName)
                 .addFetch(topicPartition.getTopic(), topicPartition.getPartition(), offset,
-                          MAX_MESSAGE_SIZE_BYTES)
+                        MAX_MESSAGE_SIZE_BYTES)
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: ??

File: src/main/java/com/pinterest/secor/common/KafkaClient.java [hunk 11]
```diff
@@ -160,16 +164,16 @@ public int getNumPartitions(String topic) {
         SimpleConsumer consumer = null;
         try {
             consumer = createConsumer(
-                mConfig.getKafkaSeedBrokerHost(),
-                mConfig.getKafkaSeedBrokerPort(),
-                "partitionLookup");
+                    mConfig.getKafkaSeedBrokerHost(),
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: ??

File: src/main/java/com/pinterest/secor/reader/MessageReader.java [hunk 12]
```diff
@@ -162,7 +162,7 @@ public Message read() {
         MessageAndMetadata<byte[], byte[]> kafkaMessage = mIterator.next();
         Message message = new Message(kafkaMessage.topic(), kafkaMessage.partition(),
                                       kafkaMessage.offset(), kafkaMessage.key(),
-                                      kafkaMessage.message());
+                                      kafkaMessage.message(), null);
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/reader/MessageReader.java]: Why you pass 'null' here instead of reading timestamp from kafkaMessage?

File: pom.xml (hunk 4 above)
Comment (tygrash) [inline pom.xml]: Then we should do this version upgrade first, because 0.8 apis do not support extraction of timestamp. I will raise separate PR for activation of 0.10 profile before timestamp changes are merged.

File: src/main/java/com/pinterest/secor/common/KafkaClient.java (hunk 6 above)
Comment (tygrash) [inline src/main/java/com/pinterest/secor/common/KafkaClient.java]: Yes, when I run mvn package, these lines are getting added. Should I keep it or remove?

**Comment by tygrash** [comment]
will reopen after kafka version upgrade! will raise PR for it too!

**Comment by HenryCaiHaiying** [comment]
Kafka version upgrade PR is landed.

**Comment by lorenasanbal** [comment]
Hi @HenryCaiHaiying,
Do you know if this feature has been included into Secor's latest release?
Thank you very much for your help in advance.
regards

**Comment by HenryCaiHaiying** [comment]
I think using kafka message timestamp is already in secor: https://github.com/pinterest/secor/pull/344

**Comment by lorenasanbal** [comment]
Thank you @HenryCaiHaiying! Is possible that this feature is only available for MessagePack format messages? How can I add the kafkaTimestamp to the output type Delimited Text Files ?
thank you again and sorry for the inconveniences

**Comment by HenryCaiHaiying** [comment]
I think this feature is also turned on for all formats, take a look at:

src/main/java/com/pinterest/secor/timestamp/Kafka10MessageTimestamp.java
